File: server/src/Movimentacoes.Domain/Movimentacoes/Movimentacao.cs
using Core.Domain.Models;
using Movimentacoes.Domain.Movimentacoes.Enums;
using Movimentacoes.Domain.Movimentacoes.Validations;
using Movimentacoes.Domain.Movimentacoes.ValueObjects;
using System;

namespace Movimentacoes.Domain.Movimentacoes
{
    public class Movimentacao : Entity<Movimentacao>
    {
        public Movimentacao() { }

        public Conta Conta { get; private set; }
        public TipoMovimentacao Tipo { get; private set; }
        public Vinculo Vinculo { get; private set; }
        public long Valor { get; private set; }
        public DateTime DataHoraCriacao { get; private set; }

        public void DefinirConta(Guid contaId, string numero, string digitoVerificador, Guid clienteId, string nomeCompleto, string cpf) =>
            Conta = Conta.Factory.CriarConta(contaId, numero, digitoVerificador, clienteId, nomeCompleto, cpf);

        public void DefinirVinculo(Guid vinculadoId, TipoVinculo tipo) =>
            Vinculo = Vinculo.Factory.CriarVinculo(vinculadoId, tipo);

        public override bool EstaConsistente()
        {
            ValidationResult = new MovimentacaoEstaConsistenteValidation(this).Validate(this);
            return ValidationResult.IsValid;
        }

        public static class Factory
        {
            public static Movimentacao CriarMovimentacao(Guid id, TipoMovimentacao tipo, long valor, DateTime dataHoraCriacao) =>
                new Movimentacao
                {
                    Id = id,
                    Tipo = tipo,
                    Valor = valor,
                    DataHoraCriacao = dataHoraCriacao
                };
        }
    }
}

File: clients/angular/src/app/models/movimentacoes/movimentacao.ts
import { TipoVinculado } from './enums/tipo-vinculado';
import { TipoMovimentacao } from './enums/tipo-movimentacao';

export class Movimentacao {
    public id: string;
    public conta: Conta;
    public tipo: TipoMovimentacao;
    public vinculo: Vinculo;
    public valorEmReais: number;
    public dataHoraCriacao: string;
}

export class Conta {
    public id: string;
    public numero: string;
    public digitoVerificador: string;
    public cliente: Cliente;
}

export class Cliente {
    public id: string;
    public nomeCompleto: string;
    public cpf: string;
}

export class Vinculo {
    public vinculadoId: string;
    public tipo: TipoVinculado;
}
